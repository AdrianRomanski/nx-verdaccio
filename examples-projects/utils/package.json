{
  "name": "@org/utils",
  "version": "0.0.1",
  "dependencies": {}
}
