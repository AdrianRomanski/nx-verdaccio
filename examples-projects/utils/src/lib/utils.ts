import { User } from '@org/models';

export function sortUser(users: User[]): User[] {
  return users.sort((a, b) => a.name.localeCompare(b.name));
}
