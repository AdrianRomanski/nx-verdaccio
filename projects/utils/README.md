# utils

This library provides utility functions.
