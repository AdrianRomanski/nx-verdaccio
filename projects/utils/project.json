{
  "name": "utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "projects/utils/src",
  "projectType": "library",
  "release": {
    "version": {
      "generatorOptions": {
        "packageRoot": "dist/{projectRoot}",
        "currentVersionResolver": "git-tag"
      }
    }
  },
  "tags": ["scope:shared", "type:util", "publishable"],
  "targets": {
    "lint": {},
    "build": {
      "executor": "@nx/esbuild:esbuild",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/projects/utils",
        "main": "projects/utils/src/index.ts",
        "tsConfig": "projects/utils/tsconfig.lib.json",
        "assets": ["projects/utils/*.md"],
        "generatePackageJson": true,
        "format": ["esm"]
      }
    },
    "unit-test": {
      "executor": "@nx/vite:test",
      "outputs": ["{options.reportsDirectory}"],
      "options": {
        "reportsDirectory": "../../coverage/projects/utils"
      }
    }
  }
}
