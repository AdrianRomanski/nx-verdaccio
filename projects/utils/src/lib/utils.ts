import {User} from "@push-based/models";

export function sortUser(users: User[]): User[] {
  return users.sort((a, b) => a.name.localeCompare(b.name));
}
