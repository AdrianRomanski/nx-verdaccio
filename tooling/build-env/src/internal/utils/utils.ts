import { mkdir } from 'node:fs/promises';
import type { TargetConfiguration } from '@nx/devkit';

export function getBuildOutput(target?: TargetConfiguration) {
  const { options } = target ?? {};
  const { outputPath } = options ?? {};
  if (!outputPath) {
    throw new Error('outputPath is required');
  }
  return outputPath;
}

export function uniquePort(): number {
  return Number((6000 + Number(Math.random() * 1000)).toFixed(0));
}
