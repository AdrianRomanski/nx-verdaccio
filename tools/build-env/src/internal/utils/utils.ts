import { mkdir } from 'node:fs/promises';

export function uniquePort(): number {
  return Number((6000 + Number(Math.random() * 1000)).toFixed(0));
}

export async function ensureDirectoryExists(baseDir: string) {
  try {
    await mkdir(baseDir, { recursive: true });
    return;
  } catch (error) {
    console.error((error as { code: string; message: string }).message);
    if ((error as { code: string }).code !== 'EEXIST') {
      throw error;
    }
  }
}
