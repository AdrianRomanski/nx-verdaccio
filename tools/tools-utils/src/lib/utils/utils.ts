import { mkdir } from 'node:fs/promises';
import { ProjectConfiguration } from 'nx/src/config/workspace-json-project-json';

export function getBuildOutputPathFromBuildTarget(
  projectConfiguration: ProjectConfiguration
) {
  const { targets } = projectConfiguration;
  const { build } = targets ?? {};
  const { options } = build ?? {};
  const { outputPath } = options ?? {};
  if (outputPath == null) {
    throw new Error('outputPath is required');
  }
  return outputPath;
}

export async function ensureDirectoryExists(baseDir: string) {
  try {
    await mkdir(baseDir, { recursive: true });
    return;
  } catch (error) {
    console.error((error as { code: string; message: string }).message);
    if ((error as { code: string }).code !== 'EEXIST') {
      throw error;
    }
  }
}
