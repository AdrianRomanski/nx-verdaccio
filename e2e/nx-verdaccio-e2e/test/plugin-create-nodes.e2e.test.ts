import type { Tree } from '@nx/devkit';
import { join } from 'node:path';
import { afterEach, expect } from 'vitest';
import {
  addJsLibToWorkspace,
  materializeTree,
  npmOptions,
  nxShowProjectJson,
  registerPluginInWorkspace,
} from '@push-based/test-nx-utils';
import { updateProjectConfiguration } from 'nx/src/generators/utils/project-configuration';

describe('nx-verdaccio plugin create-nodes-v2', () => {
  const e2eProjectName = process.env['NX_TASK_TARGET_PROJECT'];
  let tree: Tree;
  const project = 'my-lib';
  const projectE2e = `${project}-e2e`;
  const e2eProjectRoot = join('projects', projectE2e);
  const baseDir = `tmp/environments/${e2eProjectName}/__test__/create-nodes-v2`;

  beforeEach(async () => {
    tree = await addJsLibToWorkspace(project);
    tree = await addJsLibToWorkspace(projectE2e, tree);
    updateProjectConfiguration(tree, projectE2e, {
      root: e2eProjectRoot,
      projectType: 'application',
    });
  });

  afterEach(async () => {
    // await teardownTestFolder(baseDir);
  });

  it('should add package targets to library project dynamically', async () => {
    const cwd = join(baseDir, 'add-pkg-targets');
    registerPluginInWorkspace(tree, {
      plugin: '@push-based/nx-verdaccio',
      options: {
        environments: {
          targetNames: ['e2e'],
        },
      },
    });
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, project);
    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual({
      'nxv-pkg-install': expect.objectContaining({
        dependsOn: [
          {
            target: 'nxv-pkg-publish',
            params: 'forward',
          },
          {
            target: 'nxv-pkg-install',
            projects: 'dependencies',
            params: 'forward',
          },
        ],
        executor: '@push-based/nx-verdaccio:pkg-install',
      }),
      'nxv-pkg-publish': expect.objectContaining({
        dependsOn: [
          {
            params: 'forward',
            target: 'build',
          },
          {
            params: 'forward',
            projects: 'dependencies',
            target: 'nxv-pkg-publish',
          },
        ],
        executor: '@push-based/nx-verdaccio:pkg-publish',
      }),
    });

    expect(projectJson.targets).toMatchSnapshot();
  });
  it('should NOT add package targets to application project', async () => {
    const cwd = join(baseDir, 'no-pkg-targets');
    registerPluginInWorkspace(tree, {
      plugin: '@push-based/nx-verdaccio',
      options: {
        environments: {
          targetNames: ['e2e'],
        },
      },
    });
    await materializeTree(tree, cwd);

    const { projectJson } = await nxShowProjectJson(cwd, projectE2e);

    expect(projectJson.targets).toStrictEqual(
      expect.not.objectContaining({
        'nxv-pkg-install': expect.any(Object),
        'nxv-pkg-publish': expect.any(Object),
      })
    );
  });

  it('should add environment targets to project with targetName e2e dynamically', async () => {
    const cwd = join(baseDir, 'add-env-targets');
    registerPluginInWorkspace(tree, {
      plugin: '@push-based/nx-verdaccio',
      options: {
        environments: {
          targetNames: ['e2e'],
        },
      },
    });
    updateProjectConfiguration(tree, projectE2e, {
      root: e2eProjectRoot,
      projectType: 'application',
      targets: {
        e2e: {},
      },
    });
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, projectE2e);
    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual(
      expect.objectContaining({
        e2e: expect.objectContaining({
          dependsOn: [
            {
              params: 'forward',
              target: 'nxv-env-setup',
            },
          ],
        }),
        'nxv-env-bootstrap': expect.objectContaining({
          executor: '@push-based/nx-verdaccio:env-bootstrap',
          options: {
            environmentRoot: 'tmp/environments/my-lib-e2e',
          },
        }),
        'nxv-env-install': expect.objectContaining({
          dependsOn: [
            {
              params: 'forward',
              projects: 'dependencies',
              target: 'nxv-pkg-install',
            },
          ],
          executor: 'nx:noop',
          options: { environmentRoot: 'tmp/environments/my-lib-e2e' },
        }),
        'nxv-env-setup': expect.objectContaining({
          cache: false,
          executor: '@push-based/nx-verdaccio:env-setup',
          options: {
            environmentRoot: 'tmp/environments/my-lib-e2e',
          },
          outputs: [
            '{options.environmentRoot}/node_modules',
            '{options.environmentRoot}/package.json',
            '{options.environmentRoot}/.npmrc',
            '{options.environmentRoot}/package-lock.json',
          ],
        }),
        'nxv-env-verdaccio-start': expect.objectContaining({
          executor: '@nx/js:verdaccio',
          options: expect.objectContaining({
            clear: true,
            config: '.verdaccio/config.yml',
            environmentDir: 'tmp/environments/my-lib-e2e',
            port: expect.any(Number), // random port number
            projectName: 'my-lib-e2e',
            storage: 'tmp/environments/my-lib-e2e/storage',
          }),
        }),
        'nxv-env-verdaccio-stop': expect.objectContaining({
          executor: '@push-based/nx-verdaccio:kill-process',
          options: {
            filePath: 'tmp/environments/verdaccio-registry.json',
          },
        }),
      })
    );

    expect({
      ...projectJson.targets,
      ['nxv-env-verdaccio-start']: {
        ...projectJson.targets['nxv-env-verdaccio-start'],
        options: {
          ...projectJson.targets['nxv-env-verdaccio-start'].options,
          port: expect.any(Number),
        },
      },
    }).toMatchSnapshot();
  });
  it('should NOT add environment targets to project without targetName e2e', async () => {
    const cwd = join(baseDir, 'no-env-targets');
    registerPluginInWorkspace(tree, {
      plugin: '@push-based/nx-verdaccio',
      options: {
        environments: {
          targetNames: ['e2e'],
        },
      },
    });
    await materializeTree(tree, cwd);

    const { projectJson } = await nxShowProjectJson(cwd, projectE2e);

    expect(projectJson.targets).toStrictEqual(
      expect.not.objectContaining({
        'nxv-env-bootstrap': expect.any(Object),
        'nxv-env-install': expect.any(Object),
        'nxv-env-setup': expect.any(Object),
        'nxv-env-verdaccio-start': expect.any(Object),
        'nxv-env-verdaccio-stop': expect.any(Object),
      })
    );
  });
});
