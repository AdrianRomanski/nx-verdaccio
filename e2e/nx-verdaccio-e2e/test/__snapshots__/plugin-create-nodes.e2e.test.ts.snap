// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nx-verdaccio plugin create-nodes-v2 > should add environment targets to project with targetName e2e dynamically 1`] = `
{
  "e2e": {
    "configurations": {},
    "dependsOn": [
      {
        "params": "forward",
        "target": "nxv-env-setup",
      },
    ],
    "executor": "nx:noop",
    "options": {},
    "parallelism": true,
  },
  "nxv-e2e": {
    "configurations": {},
    "dependsOn": [
      {
        "params": "forward",
        "target": "e2e",
      },
    ],
    "executor": "@push-based/nx-verdaccio:env-teardown",
    "options": {},
    "parallelism": true,
  },
  "nxv-env-bootstrap": {
    "configurations": {},
    "executor": "@push-based/nx-verdaccio:env-bootstrap",
    "options": {},
    "parallelism": true,
  },
  "nxv-env-install": {
    "configurations": {},
    "dependsOn": [
      {
        "params": "forward",
        "projects": "dependencies",
        "target": "nxv-pkg-install",
      },
    ],
    "executor": "nx:run-commands",
    "options": {
      "command": "echo "dependencies installed for tmp/environments/lib-a-e2e"",
      "environmentRoot": "tmp/environments/lib-a-e2e",
    },
    "parallelism": true,
  },
  "nxv-env-setup": {
    "cache": true,
    "configurations": {},
    "executor": "@push-based/nx-verdaccio:env-setup",
    "inputs": [
      "{projectRoot}/project.json",
      {
        "runtime": "node --version",
      },
      {
        "runtime": "npm --version",
      },
      {
        "externalDependencies": [
          "verdaccio",
        ],
      },
      "^production",
    ],
    "options": {},
    "outputs": [
      "{options.environmentRoot}/.npmrc",
      "{options.environmentRoot}/package.json",
      "{options.environmentRoot}/package-lock.json",
      "{options.environmentRoot}/node_modules",
    ],
    "parallelism": true,
  },
  "nxv-env-teardown": {
    "configurations": {},
    "executor": "@push-based/nx-verdaccio:env-teardown",
    "options": {},
    "parallelism": true,
  },
  "nxv-verdaccio-start": {
    "configurations": {},
    "executor": "@nx/js:verdaccio",
    "options": {
      "clear": true,
      "config": ".verdaccio/config.yml",
      "environmentDir": "tmp/environments/lib-a-e2e",
      "port": Any<Number>,
      "projectName": "lib-a-e2e",
      "storage": "tmp/environments/lib-a-e2e/storage",
    },
    "outputs": [
      "{options.environmentRoot}/storage",
    ],
    "parallelism": true,
  },
  "nxv-verdaccio-stop": {
    "configurations": {},
    "executor": "@push-based/nx-verdaccio:kill-process",
    "options": {
      "filePath": "tmp/environments/verdaccio-registry.json",
    },
    "parallelism": true,
  },
}
`;
